### PR TITLE
Cleanup QuantityValue interface and implementation

### LINKED ISSUE #580: Units in QuantityValues are stored internally as strings
When constructing a QuantityValue, for instance with `Datamodel.makeQuantityValue(new BigDecimal("123"), myUnit)`, the unit will internally be stored using its IRI. It will then be reconstructed using `ItemIdValueImpl.fromIri(unitIRI)` when accessed via `QuantityValue.getUnitItemId()`. If the unit is accessed multiple times, the IRI will be parsed every time into an ItemIdValue, which is perhaps a bit wasteful.

Instead, the unit object should be preserved and stored directly as such.

### PR DESCRIPTION
**Pull request by wetneb**
Changes QuantityValue to use Optionals instead of nulls and stores the unit as ItemIdValue directly instead of converting back and forth with Strings.

Closes #580.

### CODE REVIEW

**Review by Tpt — COMMENTED** [review]
Thank you! I have two concerns:

* We removed all usages of `java.util.Optional` to keep compatibility with Android. `Optional` has been added to Android 7. I am not sure if it a reasonable requirement.
* I believe that Wikibase more or less allows any URI to be used in quantity units. Enforcing the URI to be a valid Wikibase item URI might create problems when parsing some Wikibase (Wikidata ?) dumps. I am not sure if it is a problem in practice. Do you have insights on this?

**Comment by robertvazan** [comment]
@Tpt I am using `Optional` in #570. It's also requested for all API methods in #327.

**Comment by Tpt** [comment]
> @Tpt I am using Optional in #570. It's also requested for all API methods in #327.

I have not thought about it. I just had a look at Android version statistics. It seems that Android 7+ has around 90% market share now so it seems reasonable to drop support for all versions below 7. It means we might also use other Java 8 APIs like Streams.

**Comment by wetneb** [comment]
> We removed all usages of java.util.Optional to keep compatibility with Android. Optional has been added to Android 7. I am not sure if it a reasonable requirement.

Arf, I had the vague intuition that we were going in the reverse direction… My bad, I shouldn't have tried to tackle two issues at once (the goal was to avoid introducing new methods which would have to be deprecated soon after when Optionals are introduced).

> I believe that Wikibase more or less allows any URI to be used in quantity units. Enforcing the URI to be a valid Wikibase item URI might create problems when parsing some Wikibase (Wikidata ?) dumps. I am not sure if it is a problem in practice. Do you have insights on this?

Oops, I should have checked that! I just followed my intuition, given by the user interface, where you can only use items from the Wikibase instance as units.

I'd be interested to see what happens when you insert a QuantityValue with an unrelated unit URI with the API. Does the edit succeed? How is it rendered afterwards, especially when trying to edit it?

Similar question for the calendar URI in TimeValues actually. And I guess the globe IRI for the geographical coordinates too, but that's not even exposed by the interface at all AFAIK.

**Comment by Tpt** [comment]
>>  We removed all usages of java.util.Optional to keep compatibility with Android. Optional has been added to Android 7. I am not sure if it a reasonable requirement.
> Arf, I had the vague intuition that we were going in the reverse direction… My bad, I shouldn't have tried to tackle two issues at once (the goal was to avoid introducing new methods which would have to be deprecated soon after when Optionals are introduced).

It's fine :-). I believe we can introduce again `Optional`s, Android 7+ is now much more widespread than 2+ years ago.

> I'd be interested to see what happens when you insert a QuantityValue with an unrelated unit URI with the API. Does the edit succeed? How is it rendered afterwards, especially when trying to edit it?

I don't know. I believe there was some usages in the data of URIs like `https://www.wikidata.org/wiki/QXXX` a few years ago but it's maybe clean now. Do you want to test it?

File: wdtk-datamodel/src/main/java/org/wikidata/wdtk/datamodel/interfaces/QuantityValue.java [hunk 1]
```diff
@@ -64,7 +79,14 @@
...
 	 * @throws IllegalArgumentException if the unit is not "1" (no unit) or a valid item IRI
+	 * @deprecated use {@link QuantityValue.getOptionalUnit} to avoid use of null
 	 */
+	@Deprecated
 	ItemIdValue getUnitItemId();
+	
+	/**
+	 * @return the unit of this quantity as an item id value.
+	 */
+	Optional<ItemIdValue> getOptionalUnitItemId();
```
Comment (robertvazan) [inline wdtk-datamodel/src/main/java/org/wikidata/wdtk/datamodel/interfaces/QuantityValue.java:94]: Just `getUnit()` would suffice.

**Review by robertvazan — COMMENTED** [review]
Sorry for the delay. I was a bit busy. I've added some comments. Optionals in parameters are the most important issue.

**Comment by wetneb** [comment]
I tried adding a quantity value with an arbitrary URL as unit and this is what I got:
```json
{
  "errors": [
    {
      "code": "modification-failed",
      "data": {
        "messages": [
          {
            "name": "wikibase-validator-not-allowed",
            "parameters": [
              "http://foo.com/bar"
            ],
            "html": {
              "*": "Illegal value: <a rel=\"nofollow\" class=\"external free\" href=\"http://foo.com/bar\">http://foo.com/bar</a>"
            }
          }
        ]
      },
      "module": "main",
      "*": "Illegal value: http://foo.com/bar"
    }
  ],
  "servedby": "mw1357",
  "*": "See https://test.wikidata.org/w/api.php for API usage. Subscribe to the mediawiki-api-announce mailing list at &lt;https://lists.wikimedia.org/mailman/listinfo/mediawiki-api-announce&gt; for notice of API deprecations and breaking changes."
}
```

That being said, when providing a syntactically valid entity URI in the Wikibase instance itself, the existence of the corrersponding entity is not checked:

image

I could create [this statement on test.wikidata.org](https://test.wikidata.org/w/index.php?title=Q106263&oldid=541127) while http://test.wikidata.org/entity/Q1062348374837438473863 obviously does not exist.

But that would not make WDTK choke anyway (since the check I introduce is syntactical only). So I think it would mostly bring WDTK in line with Wikibase, which is probably a good thing?

If you agree, I can redo a PR without the Optional business, which does add a lot more noise and is independent.

**Comment by wetneb** [comment]
Replaced by #593.

**Pull request closed without merging**